WIP: Add e2e measure tool for API coverage

oomichi opened the pull request:
Design doc: https://docs.google.com/document/d/1Upo8lTeR7gy4olD6hASTCdn8Vy7WuDwkO_o9Y5nrxhs/edit?usp=sharing

**What this PR does / why we need it**:

This patch provides a way to know which APIs of k8s are tested on e2e.
This adds the tool compares the log and Swagger spec of k8s, then outputs which APIs are not tested.
For example, 155 APIs are tested with Conformance tests on all 848 APIs from this tool measuring.

Usage:
      $ go run hack/e2e.go -- -v --test --test_args="--ginkgo.focus=\[Conformance\] -v=8" > api.log
      $ go run apicoverage.go --restlog=api.log

Output sample:
$ go run test/e2e/apicoverage/apicoverage.go --restlog=conformance-api-20170919.log
All APIs    : 848
numTested   : 155
numNotTested: 693
  numStableAPIs: 281
  numBetaAPIs  : 318
  numAlphaAPIs : 94
Untested stable APIs:
  GET /apis/storage.k8s.io/v1/storageclasses
  POST /apis/storage.k8s.io/v1/storageclasses
  DELETE /apis/storage.k8s.io/v1/storageclasses
  PUT /api/v1/proxy/namespaces/{namespace}/pods/{name}
  POST /api/v1/proxy/namespaces/{namespace}/pods/{name}
  DELETE /api/v1/proxy/namespaces/{namespace}/pods/{name}
  OPTIONS /api/v1/proxy/namespaces/{namespace}/pods/{name}


**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes #24763

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

dims commented:
/ok-to-test

oomichi commented:
/retest

oomichi commented:
/retest

jagosan commented:
/cc @mml @mbohlool

mml commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -714,6 +715,11 @@ func (r *Request) Do() Result {
 
 	var result Result
 	err := r.request(func(req *http.Request, resp *http.Response) {
+		rest_log := fmt.Sprintf("Request: %s %s\n", req.Method, req.URL)
```
I think we should use glog here rather than inventing a new logging library.

Also, this code is widely used by other clients.  We should disable verbose operation logging by default.

oomichi replied:
@mml Thanks for your review, I completely agree with you.
This was a prototype and I just wanted to get feedback to be confident for this design and direction before spending much time and power.
Many folks give me positive feedback, so I will fix these points.
Thanks again anyways :-)

oomichi replied:
I found round_tripper.go already outputs the same log, so it is fine to avoid adding this.

mml reviewed (changes requested):
Also, can you paste a sample session, including the output?  Thanks!

mbohlool commented on test/e2e/apicoverage/apicoverage.go:
```diff
@@ -0,0 +1,175 @@
+	var apisOpenapi apiArray
+
+	if len(openapi) == 0 {
+		url := "https://raw.githubusercontent.com/kubernetes/kubernetes/master/api/openapi-spec/swagger.json"
```
I suggest you don't accept a local file and parameterize the url to accept a branch instead of assuming master.

oomichi replied:
That also is a nice idea, OK let's do that :-)

mbohlool commented on test/e2e/apicoverage/apicoverage.go:
```diff
@@ -0,0 +1,175 @@
+var reOpenapi = regexp.MustCompile(`({\S+})`)
+
+func parseOpenAPI(openapi string) apiArray {
+	var decodeData interface{}
```
better to use [go-openapi/spec.Swagger](https://github.com/go-openapi/spec/blob/master/swagger.go#L30)?

oomichi replied:
@mbohlool Oh, that is a nice idea. Yeah, it would be nice to avoid re-inventing the wheel.

mbohlool commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -714,6 +715,11 @@ func (r *Request) Do() Result {
 
 	var result Result
 	err := r.request(func(req *http.Request, resp *http.Response) {
+		rest_log := fmt.Sprintf("Request: %s %s\n", req.Method, req.URL)
```
Also I think the assumption here is we always use client go in our system. While that is true right now, it may not be true later and it is not definitely a requirement for conformance. For example we may write a controller with python client and package it as part of the system.

I suggest you do it server side. We also should not do this logging all the time as it could get big and costly. We need to think on how to enable this for kubernetes instances that we want to assess for conformance. Just thinking out loud: runtime flag?!

mbohlool reviewed (commented):
Overall I like the approach and I think we can get somewhere with this. I would start with a design doc though so we can discuss (corner) cases and solidify design before implementation.

liggitt commented:
Is coverage of every verb on every resource/subresource on every API group and version a goal for e2e tests? I would have expected that to happen more in crud unit tests (and actually does for most resources via the storage tests)

mml commented:
@liggitt I don't think so.  But this tool would let us quickly know, e.g.,
"you ran the whole conformance test and no one ever called DELETE on a node
object".

Or it would let us see the delta in verbs exercised before/after a PR.  As
above, I think this will be important more for the conformance suite than
the e2e tests generally, because we have an interest in making the
definition of conformance stable.

On Fri, Sep 8, 2017 at 4:07 PM, Jordan Liggitt <notifications@github.com>
wrote:

> Is coverage of every verb on every resource/subresource on every API group
> and version a goal for e2e tests? I would have expected that to happen more
> in crud unit tests (and actually does for most resources via the storage
> tests)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/50627#issuecomment-328233742>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAMK4Sba2F2TYPk911TeQql9Y3j-6Guks5sgci0gaJpZM4O2r6g>
> .
>

oomichi commented:
@mml 

> Also, can you paste a sample session, including the output? Thanks!

Is it fine to add the sample session as commit message ? or a sample file ?

oomichi commented:
/test pull-kubernetes-verify

oomichi commented:
/test pull-kubernetes-verify

oomichi commented:
humm, I cannot find the failure reason of pull-kubernetes-verify from the log:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/50627/pull-kubernetes-verify/52717/

I0922 23:58:21.097] hack/make-rules/../../hack/verify-bazel.sh
I0922 23:58:21.097] Makefile:126: recipe for target 'verify' failed
W0922 23:58:21.198] make: *** [verify] Error 1
W0922 23:58:23.316] Traceback (most recent call last):
W0922 23:58:23.316]   File "/var/lib/jenkins/workspace/pull-kubernetes-verify/./test-infra/jenkins/../scenarios/kubernetes_verify.py", line 111, in <module>
W0922 23:58:23.316]     main(ARGS.branch, ARGS.script, ARGS.force)
W0922 23:58:23.316]   File "/var/lib/jenkins/workspace/pull-kubernetes-verify/./test-infra/jenkins/../scenarios/kubernetes_verify.py", line 95, in main
W0922 23:58:23.316]     'bash', '-c', 'cd kubernetes && %s' % script,
W0922 23:58:23.316]   File "/var/lib/jenkins/workspace/pull-kubernetes-verify/./test-infra/jenkins/../scenarios/kubernetes_verify.py", line 53, in check
W0922 23:58:23.317]     subprocess.check_call(cmd)
W0922 23:58:23.317]   File "/usr/lib/python2.7/subprocess.py", line 186, in check_call
W0922 23:58:23.317]     raise CalledProcessError(retcode, cmd)
W0922 23:58:23.317] subprocess.CalledProcessError: Command '('docker', 'run', '--rm=true', '--privileged=true', '-v', '/var/run/docker.sock:/var/run/docker.sock', '-v', '/etc/localtime:/etc/localtime:ro', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-verify/go/src/k8s.io/kubernetes:/go/src/k8s.io/kubernetes', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-verify/go/src/k8s.io/kubernetes/_artifacts:/workspace/artifacts', '-e', 'KUBE_FORCE_VERIFY_CHECKS=n', '-e', 'KUBE_VERIFY_GIT_BRANCH=master', '-e', 'REPO_DIR=/var/lib/jenkins/workspace/pull-kubernetes-verify/go/src/k8s.io/kubernetes', 'gcr.io/k8s-testimages/kubekins-test:1.8-v20170906-3a1c5ae5', 'bash', '-c', 'cd kubernetes && ./hack/jenkins/verify-dockerized.sh')' returned non-zero exit status 2
E0922 23:58:23.321] Command failed

oomichi commented:
/test pull-kubernetes-verify

oomichi commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-kops-aws

madhusudancs commented:
I am probably not the right person for this PR.

/unassign

fejta commented:
/assign

Agree with the design doc idea
@kubernetes/sig-testing-api-reviews (is this where we want design doc proposals?)
/kind design

fejta commented:
ref https://github.com/kubernetes/test-infra/issues/5231

fejta commented:
Also would this tool be a better fit for test-infra?

oomichi commented:
/retest

Commit:
Add e2e measure tool for API coverage

This patch provides a way to know which APIs of k8s are tested on e2e.
This adds the tool compares the log and Swagger spec of k8s, then
outputs which APIs are not tested.
For example, 155 APIs are tested with Conformance tests on all 848 APIs
from this tool measuring.

Usage:
 $ go run hack/e2e.go -- -v --test \
   --test_args="--ginkgo.focus=[Conformance] -v=8" > api.log
 $ go run apicoverage.go --restlog=api.log

Signed-off-by: Nikita Komarov <nikita@kmrov.ru>

fixes #24763

oomichi commented:
I have written a design doc for this measuring tool:

https://docs.google.com/document/d/1Upo8lTeR7gy4olD6hASTCdn8Vy7WuDwkO_o9Y5nrxhs/edit?usp=sharing

Happy to see feedback :)

mml commented:
@oomichi I reached out via email, but am happy to review the design doc and this PR now.  Find API coverage holes is generally a thing we want for the conformance program.

oomichi commented:
On kubernetes repo at the time 2017/10/30, 69 APIs are covered on e2e conformance test.
The number of APIs is changed because Nikita has fixed the regex. Thanks Nikita.

The detail is the following:

All APIs    : 958
numTested   : 69
numNotTested: 889
  numStableAPIs: 428
  numBetaAPIs  : 364
  numAlphaAPIs : 97

The covered APIs are:
GET /apis/policy/v1beta1/namespaces/{namespace}/poddisruptionbudgets
DELETE /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/{name}
GET /api/v1/namespaces/{namespace}/replicationcontrollers/{name}
PUT /api/v1/namespaces/{namespace}/replicationcontrollers/{name}
DELETE /api/v1/namespaces/{namespace}/replicationcontrollers/{name}
GET /apis/apps/v1beta1/namespaces/{namespace}/deployments
GET /api/v1/namespaces/{namespace}/pods/{name}/log
GET /apis/extensions/v1beta1/namespaces/{namespace}/ingresses
GET /api/v1/namespaces/{namespace}/services/{name}
DELETE /api/v1/namespaces/{namespace}/services/{name}
GET /apis/batch/v1/namespaces/{namespace}/jobs
GET /apis/settings.k8s.io/v1alpha1/namespaces/{namespace}/podpresets
GET /api/v1/namespaces/{namespace}/services/{name}/proxy/{path}
GET /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path}
GET /api/v1/namespaces/{namespace}/podtemplates
GET /api/v1/namespaces
POST /api/v1/namespaces
GET /apis/extensions/v1beta1/namespaces/{namespace}/replicasets
POST /apis/extensions/v1beta1/namespaces/{namespace}/replicasets
GET /api/v1/namespaces/{namespace}/secrets
POST /api/v1/namespaces/{namespace}/secrets
GET /apis/autoscaling/v1/namespaces/{namespace}/horizontalpodautoscalers
GET /apis/batch/v1/namespaces/{namespace}/jobs/{name}
GET /api/v1/namespaces/{namespace}/serviceaccounts
POST /api/v1/namespaces/{namespace}/serviceaccounts
GET /api/v1/namespaces/{namespace}/endpoints
GET /api/v1/namespaces/{namespace}/replicationcontrollers
POST /api/v1/namespaces/{namespace}/replicationcontrollers
GET /apis/rbac.authorization.k8s.io/v1beta1/namespaces/{namespace}/rolebindings
GET /api/v1/nodes
GET /api/v1/namespaces/{namespace}/persistentvolumeclaims
PUT /api/v1/namespaces/{namespace}/configmaps/{name}
DELETE /api/v1/namespaces/{namespace}/configmaps/{name}
GET /api/v1/namespaces/{namespace}/events
POST /api/v1/namespaces/{namespace}/pods/{name}/exec
GET /api/v1/namespaces/{namespace}/configmaps
POST /api/v1/namespaces/{namespace}/configmaps
GET /api/v1/namespaces/{namespace}/pods/{name}
PUT /api/v1/namespaces/{namespace}/pods/{name}
DELETE /api/v1/namespaces/{namespace}/pods/{name}
GET /logs/
GET /apis/apps/v1beta1/namespaces/{namespace}/statefulsets
GET /api/v1/namespaces/{namespace}/pods/{name}/proxy
GET /api/v1/namespaces/{namespace}/services
POST /api/v1/namespaces/{namespace}/services
GET /apis/rbac.authorization.k8s.io/v1beta1/namespaces/{namespace}/roles
GET /apis/extensions/v1beta1/namespaces/{namespace}/deployments
GET /apis/extensions/v1beta1/namespaces/{namespace}/deployments/{name}
GET /api/v1/namespaces/{namespace}/pods
POST /api/v1/namespaces/{namespace}/pods
GET /api/v1/namespaces/{namespace}/serviceaccounts/{name}
GET /apis/networking.k8s.io/v1/namespaces/{namespace}/networkpolicies
GET /api/v1/nodes/{name}
PUT /api/v1/nodes/{name}
PATCH /api/v1/nodes/{name}
GET /api/v1/pods
GET /api/v1/namespaces/{namespace}/resourcequotas
GET /apis/extensions/v1beta1/namespaces/{namespace}/networkpolicies
POST /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions
GET /api/v1/namespaces/{namespace}/secrets/{name}
PUT /api/v1/namespaces/{namespace}/secrets/{name}
DELETE /api/v1/namespaces/{namespace}/secrets/{name}
GET /api/v1/namespaces/{namespace}/endpoints/{name}
GET /apis/extensions/v1beta1/namespaces/{namespace}/daemonsets
GET /api/v1/namespaces/{name}
DELETE /api/v1/namespaces/{name}
GET /api/v1/nodes/{name}/proxy/{path}
GET /apis/apps/v1beta1/namespaces/{namespace}/controllerrevisions
GET /api/v1/namespaces/{namespace}/limitranges

oomichi commented:
@fejta Thanks for your work. As the discussion, I start creating this on test-infra as https://github.com/kubernetes/test-infra/pull/5865

Now we have the issue on test-infra, so could you close https://github.com/kubernetes/kubernetes/issues/24763 on kubernetes/kubernetes side to avoid duplicated work ?
